state/exchange/actions.ts

```diff
@@ -0,0 +1,382 @@
+				...r,
+				balance: '0',
+				usdBalance: r.usdBalance.toString(),
+			})),
+			totalRedeemableBalance: totalRedeemableBalance.toString(),
+		};
+	}
+);
+
+export const fetchTxProvider = createAsyncThunk<any, void, ThunkConfig>(
+	'exchange/fetchTxProvider',
+	async (_, { getState, extra: { sdk } }) => {
+		const {
+			exchange: { quoteCurrencyKey, baseCurrencyKey },
+		} = getState();
+
+		if (baseCurrencyKey && quoteCurrencyKey) {
+			const txProvider = sdk.exchange.getTxProvider(baseCurrencyKey, quoteCurrencyKey);
+
+			return txProvider;
+		}
+
+		return undefined;
+	}
+);
+
+export const fetchTransactionFee = createAsyncThunk<any, void, ThunkConfig>(
+	'exchange/fetchTransactionFee',
+	async (_, { getState, extra: { sdk } }) => {
+		const {
+			exchange: { quoteCurrencyKey, baseCurrencyKey, quoteAmount, baseAmount },
+		} = getState();
+
+		if (baseCurrencyKey && quoteCurrencyKey) {
+			const [transactionFee, feeCost] = await Promise.all([
+				sdk.exchange.getTransactionFee(quoteCurrencyKey, baseCurrencyKey, quoteAmount, baseAmount),
+				sdk.exchange.getFeeCost(quoteCurrencyKey, baseCurrencyKey, quoteAmount),
+			]);
+
+			return {
+				transactionFee: transactionFee ? transactionFee.toString() : undefined,
+				feeCost: feeCost.toString(),
+			};
+		}
+
+		return { transactionFee: undefined, feeCost: undefined };
+	}
+);
+
+export const submitExchange = createAsyncThunk<any, void, ThunkConfig>(
+	'exchange/submitExchange',
+	async (_, { getState, dispatch, extra: { sdk } }) => {
+		const {
+			exchange: { quoteCurrencyKey, baseCurrencyKey, quoteAmount, baseAmount },
+		} = getState();
+
+		if (quoteCurrencyKey && baseCurrencyKey) {
+			const hash = await sdk.exchange.handleExchange(
+				quoteCurrencyKey,
+				baseCurrencyKey,
+				quoteAmount,
+				baseAmount,
+				true
+			);
+
+			if (hash) {
+				monitorTransaction({
+					txHash: hash,
+					onTxConfirmed: () => {
+						dispatch(fetchSynthBalances());
+						dispatch(fetchNumEntries());
+						dispatch({
+							type: 'exchange/setQuoteAmount',
+							payload: { value: '' },
+						});
+						dispatch({
+							type: 'exchange/setBaseAmount',
+							payload: { value: '' },
+						});
+					},
+				});
+			}
+		}
+	}
+);
+
+export const submitRedeem = createAsyncThunk<any, void, ThunkConfig>(
+	'exchange/submitRedeem',
+	async (_, { dispatch, extra: { sdk } }) => {
+		const hash = await sdk.exchange.handleRedeem();
+
+		if (hash) {
+			monitorTransaction({
+				txHash: hash,
+				onTxConfirmed: () => {
+					dispatch(fetchSynthBalances());
+					dispatch(fetchBalances());
+				},
+			});
+		}
+	}
+);
+
+export const submitApprove = createAsyncThunk<any, void, ThunkConfig>(
+	'exchange/submitApprove',
+	async (_, { getState, dispatch, extra: { sdk } }) => {
+		const {
+			exchange: { quoteCurrencyKey, baseCurrencyKey },
+		} = getState();
+
+		if (quoteCurrencyKey && baseCurrencyKey) {
+			const hash = await sdk.exchange.handleApprove(quoteCurrencyKey, baseCurrencyKey);
```

Comment:
I'm not sure about the UX requirements here but we might want to store the transactions in state like:
```
{
  hash: string,
  type: 'approve' ... etc,
  status: 'pending' | 'complete' | 'failed'
}
```
and then this gives us the opportunity to display and handle different transaction states in our UI as well as just the notifier, thinking here like we want to disable a button / show approving before being able to perform a swap. 

And probably good practice to do this for each transaction type.